pkg/ipamd/ipamd.go

```diff
@@ -884,24 +884,24 @@ func (c *IPAMContext) tryAllocateENI(ctx context.Context) error {
 		subnet = eniCfg.Subnet
 	}
 
-	eni, err := c.awsClient.AllocENI(c.useCustomNetworking, securityGroups, subnet)
+	resourcesToAllocate := c.GetENIResourcesToAllocate()
+	eni, err := c.awsClient.AllocENI(c.useCustomNetworking, securityGroups, subnet, resourcesToAllocate)
 	if err != nil {
 		log.Errorf("Failed to increase pool size due to not able to allocate ENI %v", err)
 		ipamdErrInc("increaseIPPoolAllocENI")
-		return err
-	}
-
-	resourcesToAllocate := c.GetENIResourcesToAllocate()
-	_, err = c.awsClient.AllocIPAddresses(eni, resourcesToAllocate)
-	if err != nil {
 		log.Warnf("Failed to allocate %d IP addresses on an ENI: %v", resourcesToAllocate, err)
-		// Continue to process the allocated IP addresses
-		ipamdErrInc("increaseIPPoolAllocIPAddressesFailed")
 		if containsInsufficientCIDRsOrSubnetIPs(err) {
+			ipamdErrInc("increaseIPPoolAllocIPAddressesFailed")
 			log.Errorf("Unable to attach IPs/Prefixes for the ENI, subnet doesn't seem to have enough IPs/Prefixes. Consider using new subnet or carve a reserved range using create-subnet-cidr-reservation")
 			c.lastInsufficientCidrError = time.Now()
-			return err
 		}
+
+		return err
+	}
+	// This case is to handle an ENI not being created due to 0 IPs being needed.
+	if eni == "" {
```

Comment:
IMO we shouldn't be doing these checks, if we don't need IPs then we shouldn't even call Alloc.

But trying to understand more about the change, why are we merging the functions? Still we will end up with 2 EC2 calls isn't it?

Reply:
Nvm, we are passing the count to createNetworkInterfaceWithContext..but we should refactor the eni check..